upf/substituter.py

```diff
@@ -0,0 +1,78 @@
+# Copyright 2021 AIPlan4EU project
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from upf.expression import Expression
+import upf.environment
+from upf.walkers.identitydag import IdentityDagWalker
+import upf.walkers as walkers
+import upf.operators as op
+from upf.fnode import FNode
+from typing import List
+import typing
+from upf.type_checker import TypeChecker
+
+class Substituter(IdentityDagWalker):
+    """Performs substitution into an expression """
+    def __init__(self, env: 'upf.environment.Environment'):
+        IdentityDagWalker.__init__(self, env, True)
+        self.env = env
+        self.manager = env.expression_manager
+        self.type_checker = TypeChecker(env)
+
+    def _get_key(self, expression, **kwargs):
+        return expression
+
+    def substitute(self, expression: FNode, substitutions: typing.Dict[Expression, Expression] = {}) -> FNode:
+        """Performs substitution into the given expression.
+
+        Lets consider the examples:
+        f = a & b
+        subs = {a -> c, (c & b) -> d, (a & b) -> c}
+        substitute(f, subs) = c
+
+        f = a
+        subs = {a -> c, c -> d}
+        substitute(f, subs) = c
+
+        f = a & b
+        subs = {a -> 5, b -> c}
+        substitute(f, subs) = a & c
```

Comment:
Should be something like this, right?
```suggestion
        substitute(f, subs) = 5 & c
```

Reply:
Actually in this case, since "a" must be a boolean, assigning 5 to it is just removed from the substitution dict.

However, since we are raising an exception when this happens and not simply removing it from the subst dictionary anymore, this documentation must also be changed.

Reply:
Ah ok, yes!